### DISCUSSION TITLE
Dependency and project management files/tools

### DISCUSSION THREAD

**Discussion by PEZ** [ingydotnet/lingy#29] [General]
I notice that clj-kondo (by proxy of clojure-lsp) highlights a lot of linting problems with Lingy files. I think partly it is because there's no project file that clj-kondo finds and can use to figure out the classpath and such. So starting this discussion so that we can figure about what we can do to help clj-kondo and clojure-lsp to “understand” Lingy code better.

I think that for Babashka both deps.edn and bb.edn can be used. And maybe it is a `lingy.edn` in the future which Clojure tooling can be made to use. Or, something else... I'm quite unclear about the while subject.

For ClojurCLR there are plans to port over the tools.deps stuff. Which also reminds me that I wrote a silly poor man's dependency tool for ClojurCLR a while ago. 😄
 https://blog.agical.se/en/posts/how-to-create-a-really-simple-clojureclr-dependency-tool/

**Comment by seancorfield** [comment]
Another aspect of this discussion is procurers and coordinates. In `deps.edn`, we are used to specifying dependencies as `lib/name {coords}` where the `{coords}` identify a procurer -- a "source" for dependencies -- and a version, and `lib/name` uniquely identifies the library _within that procurer_.

* `:local/root` identifies source directories containing the library (`lib/name` doesn't matter here except as some unique label for that library in your project)
* `:git/sha` identifies a git-hosted resource and `lib/name` generally follows a convention that lets the CLI deduce the URL (`io.github.account/repo` => https://github.com/account/repo for example, but several others are supported)
* `:mvn/version` identifies a Maven-style repository (Maven Central, Clojars) that provides JAR-packaged libraries with a standard set of (XML) metadata files

Clojure and ClojureScript support the above three procurers. Cross-platform Clojure libraries are written with `.cljc` files and can use reader conditionals for platform-specific code (using `:clj`, `:cljs`, `:cljr` as discriminators so far).

The ClojureCLR version of the CLI will support `:local/root` and `:git/sha`. There is still ongoing discussion about whether/how to support `nuget` which is the "standard" way .NET packages are published/distributed.

Lingy should support `:local/root` and `:git/sha` I think? Lingy Perl might want to support CPAN somehow? CPAN could be treated as new procurer with a new keyword in the coordinates to identify it. Even if fetching/installing the dependency is not automated as part of the CLI, the ability to identify library as being needed to run a program is probably needed on each host.

**Comment by seancorfield** [comment]
The follow-on discussion to that is about portable -- multi-host -- libraries. If you have a standalone library, with no dependencies, it could be depended on via `:git/sha` and used for every host. However, if the library has dependencies -- even if those dependencies exist for multiple hosts -- the coordinates may well be different. This will likely force the library itself to be published to multiple services, each one with the appropriate `deps.edn` file for the relevant host.

HoneySQL currently supports Clojure, ClojureScript, and ClojureCLR via reader conditionals: https://github.com/seancorfield/honeysql/blob/develop/src/honey/sql.cljc#L171-L186 but has no dependencies. If it had (non-git) dependencies, it's likely it would have `:mvn/version` coordinates for Clojure/Script (or other portable libraries), and some sort of `nuget` coordinates for ClojureCLR.

It's currently an open area of research as to how this might work.
